File: widgetssdk/src/main/java/com/glia/widgets/notification/device/NotificationManager.java
package com.glia.widgets.notification.device;

import android.app.Application;
import android.app.Notification;
import android.app.NotificationChannel;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.annotation.RequiresApi;

import com.glia.widgets.R;
import com.glia.widgets.notification.NotificationFactory;
import com.glia.widgets.screensharing.MediaProjectionService;

import static com.glia.widgets.notification.NotificationFactory.SCREEN_SHARING_NOTIFICATION_ID;

public class NotificationManager implements INotificationManager {
    private final Context applicationContext;
    private final android.app.NotificationManager notificationManager;

    private Notification callNotification;

    public NotificationManager(Application application) {
        this.applicationContext = application;
        this.notificationManager = application.getSystemService(android.app.NotificationManager.class);

        createNotificationChannel();
    }

    private void createNotificationChannel() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            createCallChannel();
            createScreenSharingChannel();
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void createCallChannel() {
        if (notificationManager.getNotificationChannel(NotificationFactory.NOTIFICATION_CALL_CHANNEL_ID) == null) {
            int importance = android.app.NotificationManager.IMPORTANCE_HIGH;
            NotificationChannel notificationChannel = new NotificationChannel(NotificationFactory.NOTIFICATION_CALL_CHANNEL_ID, applicationContext.getString(R.string.notification_call_channel_name), importance);
            notificationManager.createNotificationChannel(notificationChannel);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void createScreenSharingChannel() {
        if (notificationManager.getNotificationChannel(NotificationFactory.NOTIFICATION_SCREEN_SHARING_CHANNEL_ID) == null) {
            int importance = android.app.NotificationManager.IMPORTANCE_HIGH;
            NotificationChannel notificationChannel = new NotificationChannel(NotificationFactory.NOTIFICATION_SCREEN_SHARING_CHANNEL_ID, applicationContext.getString(R.string.notification_screen_sharing_channel_name), importance);
            notificationManager.createNotificationChannel(notificationChannel);
        }

    }

    @Override
    public void showAudioCallNotification() {
        callNotification = NotificationFactory.createCallStartedNotification(applicationContext);
        notificationManager.notify(NotificationFactory.CALL_NOTIFICATION_ID, callNotification);
    }

    @Override
    public void removeCallNotification() {
        if (callNotification != null) {
            notificationManager.cancel(NotificationFactory.CALL_NOTIFICATION_ID);
        }
    }

    @Override
    public void showVideoCallNotification() {
        callNotification = NotificationFactory.createVideoCallStartedNotification(applicationContext);
        notificationManager.notify(NotificationFactory.CALL_NOTIFICATION_ID, callNotification);
    }

    @Override
    public void showScreenSharingNotification() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            applicationContext.startForegroundService(new Intent(applicationContext, MediaProjectionService.class));
        } else {
            notificationManager.notify(NotificationFactory.SCREEN_SHARING_NOTIFICATION_ID, NotificationFactory.createScreenSharingNotification(applicationContext));
        }
    }

    @Override
    public void removeScreenSharingNotification() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            applicationContext.stopService(new Intent(applicationContext, MediaProjectionService.class));
        } else {
            notificationManager.cancel(SCREEN_SHARING_NOTIFICATION_ID);
        }
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/internal/notification/device/NotificationManager.kt
package com.glia.widgets.internal.notification.device

import android.app.Application
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Intent
import android.os.Build
import androidx.annotation.StringRes
import com.glia.widgets.R
import com.glia.widgets.core.notification.NotificationRemovalService
import com.glia.widgets.helper.Logger
import com.glia.widgets.helper.TAG
import com.glia.widgets.internal.notification.NotificationFactory
import com.glia.widgets.internal.notification.areNotificationsEnabledForChannel

internal class NotificationManager(
    private val applicationContext: Application,
) : INotificationManager {
    private val notificationManager: NotificationManager by lazy {
        applicationContext.getSystemService(NotificationManager::class.java)
    }

    val notificationRemovalServiceIntent: Intent by lazy {
        Intent(applicationContext, NotificationRemovalService::class.java)
    }

    init {
        createNotificationChannels()
    }

    private fun createNotificationChannels() {
        createCallChannel()
        createScreenSharingChannel()
    }

    private fun createCallChannel() = createChannelIfDoesNotExist(
        NotificationFactory.NOTIFICATION_CALL_CHANNEL_ID,
        R.string.android_notification_audio_call_channel_name,
        /* Call notification importance should have the possible highest value,
        because it shows the ongoing audio/video streaming and it is preferable to be as high as possible in notifications list */
        NotificationManager.IMPORTANCE_HIGH
    )

    private fun createScreenSharingChannel() = createChannelIfDoesNotExist(
        NotificationFactory.NOTIFICATION_SCREEN_SHARING_CHANNEL_ID,
        R.string.android_notification_screen_sharing_channel_name,
        /* Screen sharing notification importance should have the possible highest value,
            because it affects the Media projection service initialization and streaming timing */
        NotificationManager.IMPORTANCE_HIGH
    )

    override fun createSecureMessagingChannel() = createChannelIfDoesNotExist(
        NotificationFactory.NOTIFICATION_SECURE_MESSAGING_CHANNEL_ID,
        R.string.android_notification_secure_messaging_channel_name,
        NotificationManager.IMPORTANCE_HIGH
    )

    private fun createChannelIfDoesNotExist(id: String, @StringRes nameRes: Int, importance: Int) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && notificationManager.getNotificationChannel(id) == null) {
            notificationManager.createNotificationChannel(
                NotificationChannel(id, applicationContext.getString(nameRes), importance)
            )
        }
    }

    override fun showAudioCallNotification() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            // Android APIs 31+ have built in privacy indicators for microphone and camera
            // See https://source.android.com/docs/core/permissions/privacy-indicators
            return
        }
        if (areNotificationsEnabledForChannel(NotificationFactory.NOTIFICATION_CALL_CHANNEL_ID)) {
            notificationManager.notify(
                NotificationFactory.CALL_NOTIFICATION_ID,
                NotificationFactory.createAudioCallNotification(applicationContext)
            )
        }
    }

    override fun showVideoCallNotification(isTwoWayVideo: Boolean, hasAudio: Boolean) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            // Android APIs 31+ have built in privacy indicators for microphone and camera
            // See https://source.android.com/docs/core/permissions/privacy-indicators
            return
        }
        if (areNotificationsEnabledForChannel(NotificationFactory.NOTIFICATION_CALL_CHANNEL_ID)) {
            notificationManager.notify(
                NotificationFactory.CALL_NOTIFICATION_ID,
                NotificationFactory.createVideoCallNotification(applicationContext, isTwoWayVideo, hasAudio)
            )
        }
    }

    override fun startNotificationRemovalService() {
        // If this service is not already running, it will be instantiated and started (creating a process for it if needed);
        // if it is running then it remains running.
        try {
            applicationContext.startService(notificationRemovalServiceIntent)
        } catch (error: Throwable) {
            // Above code is known to throw an error if app is in background
            Logger.e(TAG, "Failed to launch 'NotificationRemovalService'", error)
        }
    }

    override fun removeCallNotification() {
        notificationManager.cancel(NotificationFactory.CALL_NOTIFICATION_ID)
        applicationContext.stopService(notificationRemovalServiceIntent)
    }

    /**
     * Displays notification informing the user that screen sharing is active.
     */
    override fun showScreenSharingNotification() = notificationManager.notify(
        NotificationFactory.SCREEN_SHARING_NOTIFICATION_ID,
        NotificationFactory.createScreenSharingNotification(applicationContext)
    )

    /**
     * Removes the screen sharing notification.
     */
    override fun removeScreenSharingNotification() = notificationManager.cancel(NotificationFactory.SCREEN_SHARING_NOTIFICATION_ID)

    override fun showSecureMessageNotification(content: String, contentIntent: PendingIntent) {
        if (areNotificationsEnabledForChannel(NotificationFactory.NOTIFICATION_SECURE_MESSAGING_CHANNEL_ID)) {
            notificationManager.notify(
                NotificationFactory.SECURE_MESSAGING_NOTIFICATION_ID,
                NotificationFactory.createSecureMessagingNotification(
                    applicationContext,
                    content,
                    contentIntent
                )
            )
        }
    }

    private fun areNotificationsEnabledForChannel(id: String) = applicationContext.areNotificationsEnabledForChannel(id)
}
